core/client/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java

```diff
@@ -0,0 +1,443 @@
+
+import alluxio.Configuration;
+import alluxio.Constants;
+import alluxio.PropertyKey;
+import alluxio.client.block.BlockStoreContext;
+import alluxio.network.protocol.RPCMessageDecoder;
+import alluxio.network.protocol.RPCProtoMessage;
+import alluxio.network.protocol.Status;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.DataNettyBufferV2;
+import alluxio.proto.dataserver.Protocol;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.protobuf.MessageLite;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelPipeline;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A netty packet writer that streams a full block or file to a netty data server.
+ *
+ * Protocol:
+ * 1. The client streams packets (start from pos 0) to the server. The client pauses if the client
+ *    buffer is full, resumes if the buffer is not full.
+ * 2. The server reads packets from the channel and writes them to the block worker. See the server
+ *    side implementation for details.
+ * 3. When all the packets are sent, the client closes the reader by sending an empty packet to
+ *    the server to signify the end of the block. The client must wait the response from the server
+ *    to make sure everything has been written to the block worker.
+ * 4. To make it simple to handle errors, the channel is closed if any error occurs.
+ */
+@NotThreadSafe
+public class NettyPacketWriter implements PacketWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final long PACKET_SIZE =
+      Configuration.getBytes(PropertyKey.USER_NETWORK_NETTY_WRITER_PACKET_SIZE_BYTES);
+  private static final int MAX_PACKETS_IN_FLIGHT =
+      Configuration.getInt(PropertyKey.USER_NETWORK_NETTY_WRITER_BUFFER_SIZE_PACKETS);
+  private static final long WRITE_TIMEOUT_MS =
+      Configuration.getLong(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+
+  private final Channel mChannel;
+  private final InetSocketAddress mAddress;
+  private final long mId;
+  private final long mSessionId;
+  private final Protocol.RequestType mRequestType;
+  private final long mLength;
+
+  private boolean mClosed = false;
+
+  private ReentrantLock mLock = new ReentrantLock();
+  /** The next pos to write to the channel. */
+  @GuardedBy("mLock")
+  private long mPosToWrite = 0;
```

Comment:
Everything before mPosToWrite are flushed to the network.

Everything before mPosToQueue are either flushed or pushed to the buffer.